pydantic_ai_slim/pydantic_ai/models/_tool_choice.py

```diff
@@ -0,0 +1,142 @@
+        return 'auto' if model_request_parameters.allow_text_output else 'required'
+
+    # none / []: disable function tools, but output tools may still exist
+    if tool_choice in ('none', []):
+        output_tool_names = [t.name for t in model_request_parameters.output_tools]
+
+        if output_tool_names:
+            # If direct output is allowed, output-tools-only behaves like "auto";
+            # otherwise if function tools exist, enforce "required" to avoid silent dead-ends.
+            if allow_direct_output:
+                mode: Literal['auto', 'required'] = 'auto'
+            elif model_request_parameters.function_tools:
+                mode = 'required'
+            else:
+                return 'required'  # only output tools exist and direct output isn't allowed
+
+            _warn(
+                f"tool_choice='none' but output tools {output_tool_names} are defined - "
+                f'defaulting to tool_choice={mode!r} for output tools only'
+            )
+            return (output_tool_names, mode)
+
+        if allow_direct_output:
+            return 'none'
+
+        # pragma: no cover
+        assert False, 'Either output_tools or allow_text_output/allow_image_output must be set'
+
+    # required (only function tools allowed)
+    if tool_choice == 'required':
+        if model_request_parameters.output_tools:
+            raise UserError(
+                "`tool_choice='required'` is incompatible with output types. "
+                'Use `ToolsPlusOutput` to specify function tools while allowing structured output.'
+            )
+        if not model_request_parameters.function_tools:
+            raise UserError(
+                '`tool_choice` was set to "required", but no function tools are defined. '
+                'Please define function tools or change `tool_choice` to "auto" or "none".'
+            )
+        return 'required'
+
+    # list[str]: required, restricted to these tools
+    if isinstance(tool_choice, list):
+        # stable order, unique
+        chosen = list(dict.fromkeys(tool_choice))
+        chosen_set = set(chosen)
+
+        if chosen_set == set(tool_defs_keys):
+            return 'required'
+
+        _invalid(chosen_set, set(tool_defs_keys), available_label='Available tools')
+        return (chosen, 'required')
+
+    # ToolsPlusOutput: specific function tools + all output tools
+    if isinstance(tool_choice, ToolsPlusOutput):
+        output_tool_names = [t.name for t in model_request_parameters.output_tools]
+
+        # stable order, unique
+        if not tool_choice.function_tools:
+            _warn("ToolsPlusOutput with empty function_tools - defaulting to 'none'")
+            if output_tool_names:
+                return (output_tool_names, 'auto' if allow_direct_output else 'required')
+            return 'none'
+
+        if not output_tool_names:
+            _warn('ToolsPlusOutput used but no output tools exist - defaulting to list[str] behavior')
```

Comment:
`ToolsPlusOutput` should also allow text/image output, if that's allowed by the output type/model params right? So then we shouldn't have a warning here, and we should send return `(allowed_tools, auto)`